rust/src/main.rs

```diff
@@ -0,0 +1,339 @@
+#![feature(test)]
+#![deny(unsafe_code)]
+
+extern crate itertools;
+extern crate mersenne_twister;
+extern crate num;
+extern crate rand;
+extern crate test;
+
+// Use BLAS directly
+#[cfg(feature = "direct_blas")]
+extern crate cblas;
+
+#[cfg(feature = "direct_blas")]
+extern crate blas_src;
+
+// Use ndarray (with BLAS implementation)
+#[cfg(not(feature = "direct_blas"))]
+#[macro_use(s)]
+extern crate ndarray;
+
+use std::time::{Duration, Instant};
+use std::u32;
+use std::fs::OpenOptions;
+use std::io::{BufWriter, Write};
+
+use test::black_box;
+use num::complex::Complex64;
+use rand::Rng;
+
+mod util;
+use util::{gen_rng, myrand};
+
+#[cfg(feature = "direct_blas")]
+mod direct_blas;
+#[cfg(feature = "direct_blas")]
+use direct_blas::{randmatstat, randmatmul, check_randmatmul};
+
+#[cfg(not(feature = "direct_blas"))]
+use ndarray::Array2;
+#[cfg(not(feature = "direct_blas"))]
+use util::fill_rand;
+#[cfg(not(feature = "direct_blas"))]
+use num::Zero;
+
+const NITER: u32 = 5;
+
+#[cfg(not(feature = "direct_blas"))]
+fn nrand<R: Rng>(shape: (usize, usize), rng: &mut R) -> Array2<f64> {
+    let mut m = Array2::zeros(shape);
+    fill_rand(&mut m, rng);
+    m
+}
+
+fn fib(n: i32) -> i32 {
+    let n = black_box(n); // prevent over-optimization
```

Comment:
What's going on here?

Reply:
That is to prevent the compiler from doing invariant code motion: it would recognize that the function had already been called with the same arguments in a previous step, and choose to memoize the output, which would distort the whole benchmark. `black_box` hints LLVM to ignore about the "nature" of the value, thus forcing it to always calculate `fib`.

Reply:
`#[inline(never)]` might work as well.

Reply:
@vks I'm afraid not, just tried that too. What makes the compiler perform invariant code motion is the perception that the pure function is being called with the exact same argument value across the multiple iterations, and so the compiler can just move it out of the loop. This is orthogonal to whether the function is inlined.